Support anchors in the hash history

just-boris opened the pull request:
Now `location.hash` is ignored when `createHashHistory` is used.

This fix makes the behavior of `createHashHistory` more compatible with `createBrowserHistory`.

taion commented:
Why? They're not going to do anything. You also didn't update `getCurrentLocation` either.

just-boris commented:
Thanks for noting this, I fixed `getCurrentLocation` too.

I am going to use it for anchorScroll behavior. I implemented this in my app, and now it works nice, except that hash does not appear in location, so I can't share a link for someone else.

taion commented on modules/DOMUtils.js:
```diff
@@ -17,7 +17,7 @@ export function removeEventListener(node, event, listener) {
   // We can't use window.location.hash here because it's not
   // consistent across browsers - Firefox will pre-decode it!
-  return window.location.href.split('#')[1] || ''
+  return window.location.href.split('#').slice(1).join('#') || ''
```
Use `indexOf` and `substr` to avoid extra work with arrays. It doesn't lose at all on readability.

taion commented:
Thanks. One more nit. I'm going to hold off on merging - code looks fine but gonna see what others think about adding this feature.

just-boris commented:
Done. 
Okay, I will be wait

taion commented:
Per http://stackoverflow.com/questions/10850781/multiple-hash-signs-in-url, `#` is not actually legal in a URL hash. Interesting - do you think it'd be okay to escape it, or would it make your URLs unnecessarily ugly?

just-boris commented:
Ok, that makes sense. I checked that other routing library, like `Backbone.router` does this too.
I am going to keep updated `getHashPath()` function to support manually typed URLs, also I will add escaping into `createHref ` and unescaping in the `getCurrentLocation `

I don't have time to do it now but I'll do it during next week.

taion commented:
I'm struggling to think of a good way to do this, though - we don't do any URI encoding/decoding elsewhere in this library. Maybe this is just okay as-is, since it seems to work properly on all the browsers we care about anyway.

just-boris commented:
Yes, that should work fine as it is now in all supported browsers. If something will go wrong, tests will show it.
So, should I update my PR?

taion commented:
What would there to be update if so?

just-boris commented:
If we are going to escape, it should be added in the `createHref ` to get escaped links, and `getCurrentLocation` to parse them back.

It will be something like this
```diff
function createHref(path) {
+    // '#' is not a valid symbol in the hash according to RFC-3986    
+    path = path.replace('#', '%23')
      return '#' + history.createHref(path)
}
```

taion commented:
I'm not sure we should escape. We're not otherwise escaping anything else in the hash string, so I'm concerned about inconsistency.

just-boris commented:
Ok, so I leave this as it is now.
Is there something else blocking merge of this?

taion commented:
I want @mjackson to take a look.

taion commented:
Ping @mjackson.

mjackson commented:
I'm not convinced this is necessary. Every day hash history becomes less and less relevant. Also, the double hash results in an invalid URL.

taion commented:
@mjackson I think for public-facing sites, using hash history these days is an anti-pattern. However, it's really convenient to have hash history for setting up one-shot things for e.g. internal use. I use this in a couple of places for internal-facing tools where my front end is purely just a bunch of static files, i.e. when I just want a super-minimal hacky front end to some API I'm building out or something.

just-boris commented:
@mjackson I am agree with @taion, I am using hash history on my project on Github Pages. I can't change server-side logic there. So  I am using my fork to do client-only routing

taion commented:
@just-boris For GH pages, the preferred solution is to render everything statically, then upload the static results to GH. This is what we do for e.g. React-Bootstrap. It's not too bad.

just-boris commented:
@taion I am not talking about a static site with React. I have a SPA that renders result from Github API. Here is it: http://catatron.com/revive/
Now I don't see any solution for me rather than this PR.

taion commented:
Oh that is quite cool. I think this broadly makes sense to me – though note that we don't currently have scroll behavior management for anchors yet anyway.

just-boris commented:
@taion yes, I made it up on my side. It doesn't require to change anything in the core library.
Source code is [here](https://github.com/just-boris/revive/blob/master/src/application/util/useAnchorScroll.js) I can share it if someone will be interested, but now we have to finish this first.

SimenB commented:
Any news here?

mjackson commented:
I believe this should work on master, as of f12d632691ebef652d0cf5dd67a033e87746fc4a. We'll release a v3 beta soon (hopefully today) that you can test.

just-boris commented:
All right, the current master should work as expected.

So, closing this PR and waiting for  3.0 release then. According to changelog, it seems that it will not be difficult to upgrade
